lib/std/fs.zig

```diff
@@ -2101,17 +2105,33 @@ pub const Dir = struct {
         BadPathName,
     } || os.UnexpectedError;
 
+    pub const DeleteTreeOptions = struct {
+        /// After successfully deleting all of a directories children, it is possible for
+        /// the directory to not actually be empty (either from a new child being added while
+        /// iterating or a child being in a `DELETE_PENDING` state on Windows). In this scenario,
+        /// the directory will be iterated again and then will be attempted to be deleted again.
+        /// This provides a limit to the total number of such retries.
+        /// In `deleteTree`, the retry limit is per-directory.
+        /// In `deleteTreeMinStackSize`, the retry limit is function-wide.
```

Comment:
Thinking about it now, it would be possible to get the same per-directory behavior in `deleteTreeMinStackSize` by resetting the `retries_remaining` after a successful delete of a directory. Can't make the change now, but I will when I have the chance.

Reply:
I think we need per nesting level counters for each dir and do DFS + get children count on each new dir exploring to detect other processes making future modifications in the directory. Or, complete block other processes from accessing dirs.
Consider the following simplified cases 1,2:
```
dir1/
dir1/dir1/file1
dir1/dir1/file2
dir1/dir2/file1
dir2/dir1/file1
dir2/dir2/file2
```
- Process A deleting and B creating directories/files.
- case 1: A has deleted `dir1/dir1/file1` and continues with dir1/dir1/file2`, B recreates `dir1/dir1/file1` (could be mitigated by checking mtime of dir, if that would be reliable from Kernels).
- case 2: A has deleted `dir1/dir1/` and continues with dir1/dir2/file1`, B recreates `dir1/dir1/` with files `dir1/dir1/file1` and `dir1/dir1/file2`.

The problem boils down to enforcing the invariant, that upon each visiting of the directory the number of files must decrease and **remain decreased**.

The question to implement this would be: Do all Kernels with file systems have a system call to atomically + reliably report the number of files and directories in a directory handle? For obvious reasons, this would also have a performance impact (1 syscall for each directory).

Reply:
I think I see why that might be desirable--to make it more likely that the tree actually gets deleted even if some other process is creating files within the tree--but I don't see it as a must-have. The implementation in this PR is only really intending on being a fail-safe in avoiding the infinite loop. It probably will fail to delete directories if e.g. some other process is constantly creating new files but that seems okay for now at least.

Unless I'm misunderstanding and you're pointing out that this implementation wouldn't actually avoid the infinite loop in some scenario.